### PR TITLE
Add WAL file archiving.

### PR DESCRIPTION
**Pull request by almathew**
I'm starting to feel pretty strongly that we should make this explicit opt-in (probably by short-circuiting archive.sh if an environment variable isn't set?), given how large the archives could get, but I'm open to opinions. 

@UserNotFound and I will be walking through the recover process + docs I have for that on Monday.

CC @krallin

### CODE REVIEW

File: bin/archive-cleanup.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+find $ARCHIVE_DIRECTORY/* -mtime +2 -exec rm {} \;
```
Comment (krallin) [inline bin/archive-cleanup.sh]: I'm concerned using `rm` next to an unquoted variable :smile: 

Do we even need a glob here? Wouldn't `find "$ARCHIVE_DIRECTORY"` do the trick?

In any case; do  `set -o nounset` here.

File: bin/archive.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+set -e
```
Comment (krallin) [inline bin/archive.sh]: let's `set -o nounset` as well

File: bin/archive.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+PATH_TO_FILE=$1
```
Comment (krallin) [inline bin/archive.sh]:
```suggestion
PATH_TO_FILE="$1"
```

File: bin/archive.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+WAL_FILE_NAME=$2
```
Comment (krallin) [inline bin/archive.sh]:
```suggestion
WAL_FILE_NAME="$2"
```

File: bin/archive.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+notify () {
+  sudo -u root /usr/bin/pagerduty-notify.sh
+}
+
+trap 'notify' ERR
+
+test ! -f $ARCHIVE_DIRECTORY/$WAL_FILE_NAME
```
Comment (krallin) [inline bin/archive.sh]:
```suggestion
test ! -f "${ARCHIVE_DIRECTORY}/${WAL_FILE_NAME}"
```

File: bin/archive.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+cp $DATA_DIRECTORY/$PATH_TO_FILE $ARCHIVE_DIRECTORY/$WAL_FILE_NAME
```
Comment (krallin) [inline bin/archive.sh]:
```suggestion
cp "${DATA_DIRECTORY}/${PATH_TO_FILE}" "${ARCHIVE_DIRECTORY}/${WAL_FILE_NAME}"
```

File: bin/run-database.sh [hunk 7]
```diff
@@ -67,14 +67,32 @@ function pg_init_conf () {
...
+}
+
+function pg_init_pagerduty_notify () {
+  cat /usr/bin/pagerduty-notify.template \
+    | sed "s:__PAGERDUTY_INCIDENT_KEY__:${PAGERDUTY_INCIDENT_KEY}:g" \
+    | sed "s:__PAGERDUTY_IDENTIFIER__:${PAGERDUTY_IDENTIFIER}:g" \
+    | sed "s:__PAGERDUTY_INTEGRATION_KEY__:${PAGERDUTY_INTEGRATION_KEY}:g" \
+    > /usr/bin/pagerduty-notify.sh
+
+  chmod 700 /usr/bin/pagerduty-notify.sh
```
Comment (krallin) [inline bin/run-database.sh]: Maybe `chown` it to make sure this file is indeed owned by root as well (yes, it'll always be the case, but if we ever change the user as which this script runs, then it'll fail to warn us).

File: bin/archive.sh [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+/bin/bash archive-cleanup.sh
```
Comment (krallin) [inline bin/archive.sh]: Those all have a shebang line - you might consider just making them executable and skipping the `/bin/bash`.

File: templates/etc/postgresql/PG_VERSION/main/postgresql.conf.template [hunk 9]
```diff
@@ -30,6 +30,8 @@ shared_buffers = 128MB
 # WRITE AHEAD LOG
 #------------------------------------------------------------------------------
 
+archive_mode = on
+archive_command = '/bin/bash /usr/bin/archive.sh "%p" "%f"'
```
Comment (krallin) [inline templates/etc/postgresql/PG_VERSION/main/postgresql.conf.template]: See comment above about treating this as an executable

File: templates/usr/bin/pagerduty-notify.template [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+EVENT_ENDPOINT="https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+
+wget ${EVENT_ENDPOINT} -o /dev/null --header='Content-Type: application/json' \
```
Comment (krallin) [inline templates/usr/bin/pagerduty-notify.template]:
```suggestion
wget "$EVENT_ENDPOINT" -o /dev/null --header='Content-Type: application/json' \
```

**Comment by krallin** [comment]
> I'm starting to feel pretty strongly that we should make this explicit opt-in (probably by short-circuiting archive.sh if an environment variable isn't set?), given how large the archives could get, but I'm open to opinions.

I'm not a big fan of making this opt-in — I feel like most customers wouldn't realize they want this until it's (unfortunately) too late. 

Do you have an estimate of how big those archives might get? Perhaps we could try that out for ourselves (e.g. using the Goldenboy and Papabear DBs, which are respectively web-sized and warehouse-sized)? After all, we only archive 2 days of data on-disk (the rest is indirectly archived through snapshots, but not on-disk). 

> @UserNotFound and I will be walking through the recover process + docs I have for that on Monday.

Do you have docs for this?

**Comment by almathew** [comment]
I believe all the feedback is resolved at this point.

> Do you have an estimate of how big those archives might get?

I looked up one extreme example I knew of - they're generating 1 GB of WAL files every 100 minutes, so for 2 days they'd be around 30 GB (this is for a 300 GB database).

Still, I can see the logic behind your argument to turn it on for everyone. I expect we'll probably see an increase in disk usage warnings for PG databases for a little while - maybe we can proactively explain why that's changing?

> Do you have docs for this?

Opened PRs for both the docs and sweetness. 

One question I had remaining is how we wanted to handle the no-op images. It seems worth creating those before they're needed - one less thing to worry about during the recovery process - but I'm not sure what the best way to do that is. Making the `CMD` a variable, and then adding additional "versions" is the first way that comes to mind, but that also seems like a lot of extra stuff just to create an image that does (almost) nothing. Is there a better way?

**Pull request closed without merging**